core/domain/exp_jobs_one_off_test.py

```diff
@@ -939,6 +1057,176 @@ def test_no_action_is_performed_for_deleted_exploration(self):
             self, exp_jobs_one_off.MathExpressionValidationOneOffJob)
 
 
+class MathExpressionUpgradationOneOffJobTests(test_utils.GenericTestBase):
+
+    ALBERT_EMAIL = 'albert@example.com'
+    ALBERT_NAME = 'albert'
+
+    VALID_EXP_ID = 'exp_id0'
+    NEW_EXP_ID = 'exp_id1'
+    EXP_TITLE = 'title'
+
+    def setUp(self):
+        super(MathExpressionUpgradationOneOffJobTests, self).setUp()
+
+        # Setup user who will own the test explorations.
+        self.signup(self.ALBERT_EMAIL, self.ALBERT_NAME)
+        self.albert_id = self.get_user_id_from_email(self.ALBERT_EMAIL)
+        self.process_and_flush_pending_tasks()
+
+    def test_explorations_are_properly_upgraded(self):
```

Comment:
Write more tests. Cover all corner and unexpected cases. Ensure that the job is idempotent if something goes wrong midway.

Reply:
Done.

Reply:
I've modified the `test_explorations_with_invalid_inputs_are_not_upgraded` method to replicate the behavio that you've mentioned. Does that look alright?

Reply:
Yes, thanks. I've added a comment but in general it looks good!